<center>
<img src="../images/fscampus_small2.png" width="1200"/>
</center>

<center>

# Investments

***Finance 2 - BFIN***

**Dr. Omer Cayirli**

Lecturer in Empirical Finance

omer_cayirli@uncbusiness.net
</center>

---



## Lecture 04
---

### Outline
*   Portfolio Theory and Practice I
    *   Risk, Return, and the Historical Record
    *   Capital Allocation to Risky Assets

---



### Return on Investments
*   Total Return (Holding Period Return)
    *   The total return realized on an investment in a given period and has two components:
        *   Capital gains in the relevant period,
        *   Income generated during the period (Dividend, coupon, rent, etc.)

$$TR = \frac{P_1 - P_0 + I}{P_0}$$

| TR  | Total return                               |
| :-- | :----------------------------------------- |
| P1  | The price of the asset at the end of the period    |
| P0  | The price of the asset at the beginning of the period |
| I   | Income generated during the holding period |

*   Total return calculations do not take the return on reinvestment of the income earned during the period into account.
*   The holding period return is not a standardized value in terms of time.

---



### Return on Investments

$\text{Holding Period Return} = \text{Capital Yield} + \text{Dividend/Income Yield}$

Expected Ending Price = \$110

Beginning Price = \$100

Expected Dividend = \$4

$HPR = \frac{E(P_1) - P_0 + E(D_1)}{P_0} = \frac{\$110 - \$100 + \$4}{\$100} = \frac{\$110 - \$100}{\$100} + \frac{\$4}{\$100} = 10\% + 4\% = 14\%$

---



### Return on Investments

$$
TR_{\text{annual}}=(1+TR)^{1/N}-1
$$

An investor bought shares of Alpha Corp. for USD 52 per share on April 1, 2024. The investor received a dividend of USD 1.80 per share and sold the shares for USD 58 on January 1, 2026. Calculate the holding-period return and the annualized total return.

$$
TR=\frac{P_1-P_0+D}{P_0}
=\frac{58-52+1.80}{52}
=0.15=15.00\%
$$

The holding period is 21 months, so $N=21/12=1.75$ years.

$$
TR_{\text{annual}}=(1+0.15)^{1/1.75}-1
=0.0831=8.31\%
$$

---

### Return on Investments

On July 1, 2023, Investor Beta purchased a government bond with a face value of €1,000 for €960. The bond pays €25 every June 30 and December 31. Beta sold the bond for €985 immediately after receiving the December 31, 2024 coupon. Calculate the holding-period return and the annualized total return.

Beta received three coupon payments during the 18-month holding period:

$$
C=3(25)=75
$$

$$
TR=\frac{P_1-P_0+C}{P_0}
=\frac{985-960+75}{960}
=0.1042=10.42\%
$$

The holding period is 18 months, so $N=1.5$ years.

$$
TR_{\text{annual}}=(1+0.1041667)^{1/1.5}-1
=0.0683=6.83\%
$$

---

### Return on Investments: Expected Return

The expected return is the probability-weighted average return across possible states:

$$
E(r)=\sum_s p(s)r(s).
$$

A share of Gamma Corp. costs USD 50 today. Its possible prices and dividends one year from now are:

| Market state | Probability | Price $P_1$ | Dividend $D_1$ | Return $r(s)$ | $p(s)r(s)$ |
|---|---:|---:|---:|---:|---:|
| Boom | 10% | 68.00 | 2.00 | 40.00% | 4.00% |
| Growth | 20% | 61.00 | 1.50 | 25.00% | 5.00% |
| Normal | 40% | 54.00 | 1.00 | 10.00% | 4.00% |
| Recession | 20% | 47.00 | 0.50 | −5.00% | −1.00% |
| Depression | 10% | 40.00 | 0.00 | −20.00% | −2.00% |
| Total | 100% |  |  |  | 10.00% |

For example, the return in a boom is

$$
r(\text{Boom})=\frac{68-50+2}{50}=0.40=40.00\%.
$$

Therefore,

$$
\begin{aligned}
E(r)
&=0.10(0.40)+0.20(0.25)+0.40(0.10)\\
&\quad+0.20(-0.05)+0.10(-0.20)\\
&=0.10=10.00\%.
\end{aligned}
$$

---

### Return on Investments: Excess Return

Excess return is the return on a risky asset minus the risk-free rate:

$$
r_{\text{excess}}(s)=r(s)-r_f.
$$

Continue with the Gamma Corp. example from cell 8 and assume a one-year risk-free rate of 4%.

| Market state | Probability | Gamma return | Excess return | $p(s)[r(s)-r_f]$ |
|---|---:|---:|---:|---:|
| Boom | 10% | 40.00% | 36.00% | 3.60% |
| Growth | 20% | 25.00% | 21.00% | 4.20% |
| Normal | 40% | 10.00% | 6.00% | 2.40% |
| Recession | 20% | −5.00% | −9.00% | −1.80% |
| Depression | 10% | −20.00% | −24.00% | −2.40% |
| Total | 100% |  |  | 6.00% |

For example, in a boom,

$$
r_{\text{excess}}(\text{Boom})=0.40-0.04=0.36.
$$

The expected excess return is

$$
\begin{aligned}
E(r-r_f)
&=0.10(0.36)+0.20(0.21)+0.40(0.06)\\
&\quad+0.20(-0.09)+0.10(-0.24)\\
&=0.06=6.00\%.
\end{aligned}
$$

Equivalently,

$$
E(r-r_f)=E(r)-r_f=10.00\%-4.00\%=6.00\%.
$$

---

### Risk in Investments
*   Two of the fundamental questions of investments:
    *   What is risk and how can we measure it?
    *   How much additional return should we earn for bearing an additional unit of risk?

---

### Risk in Investments
*   One of the main functions of financial markets is to enable economic agents to transfer their risks to willing counterparties or to assume the risks of others.
*   Financial risks are generally associated with uncertainty about the outcome of an investment.
*   Accordingly, volatility (variance) in returns is the main risk criterion.
*   Risk measures are not limited to volatility:
    *   [VaR (Value at Risk)](https://pages.stern.nyu.edu/~adamodar/pdfiles/papers/VAR.pdf)
        *   Measures the potential loss in value of a risky asset or portfolio over a defined period at a given confidence level.
    *   [Expected Shortfall or CVaR (Conditional Value at Risk)](https://mayerkrebs.com/what-is-expected-shortfall-cvar/)
        *   The expected loss given that the loss is greater than or equal to the VaR.
        *   Quantifies the losses that might be encountered in the tail.
    *   [Semivariance](https://fastercapital.com/content/Downside-deviation--Navigating-Downside-Deviation-through-Semivariance.html)
        *   Measures the dispersion of all observations that fall below the mean or target value of a set of data.
    *   Sortino Ratio
        *   Measures average return above a target per unit of downside deviation
        *   Downside deviation is the square root of target semivariance

---



### Risk in Investments
*   Modern portfolio theory and asset pricing models often begin with the assumption of a normal distribution of returns.
*   The normality assumption ensures that the standard deviation is an exact measure of risk.
*   A normal distribution;
    *   Fully defined by its mean ($\mu$) and variance ($\sigma^2$).
    *   The mean of the standard normal distribution is 0 and the variance is 1.
    *   The mean or expected value ($\mu$) of the distribution is known as the location parameter.
    *   The standard deviation ($\sigma$) of the distribution is known as the scale parameter.

<center>
<img src="../images/slide_4/distribution.png" width="1080"/>
</center>

---



In [1]:
# If needed, uncomment and run once:
# %pip install -U yfinance curl_cffi pandas numpy matplotlib scipy ipywidgets

import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import ipywidgets as widgets
from scipy.stats import norm, skew, kurtosis
from IPython.display import display, HTML, clear_output
import datetime

# --- 1. Data Download and Preparation ---
tickers = ['^SP500TR', 'QQQ', 'EEM', 'GC=F', 'AMD', 'AMZN', 'F', 'JPM', 'NVDA', 'PFE', 'TGT', 'XOM']
ticker_map = {
    '^SP500TR': 'SP500', 'QQQ': 'NDQ100', 'EEM': 'EEM', 'GC=F': 'Gold Futures',
    'AMD': 'AMD', 'AMZN': 'AMZN', 'F': 'FORD', 'JPM': 'JPM',
    'NVDA': 'NVDA', 'PFE': 'PFE', 'TGT': 'TGT', 'XOM': 'XOM'
}

default_start_date = datetime.date(1999, 12, 31)
default_through_date = datetime.date(2025, 12, 31)

def run_distribution_analysis(start_date, through_date):
    if start_date is None or through_date is None:
        print('Select both dates.')
        return
    if start_date >= through_date:
        print('The start date must be earlier than the through-date.')
        return
    end_date = through_date + datetime.timedelta(days=1)  # yfinance end is exclusive

    adj_close_df = pd.DataFrame()
    print(f"Downloading data for: {list(ticker_map.values())} from {start_date} through {through_date}...")

    try:
        data_downloaded = yf.download(
            tickers,
            start=start_date,
            end=end_date,
            interval='1mo',
            group_by='ticker',
            progress=False,
            auto_adjust=True,
            actions=False
        )
        if data_downloaded.empty:
            raise ValueError("yf.download returned an empty DataFrame.")

        # With auto_adjust=True, Close is the adjusted-close series.
        df_list = []
        for ticker in tickers:
            if ticker in data_downloaded.columns.get_level_values(0):
                ticker_data = data_downloaded[ticker]
                if 'Close' in ticker_data.columns and not ticker_data['Close'].isnull().all():
                    df_list.append(ticker_data['Close'].rename(ticker_map[ticker]))

        if not df_list:
            raise ValueError("No valid adjusted-close data extracted for any ticker.")
    
        adj_close_df = pd.concat(df_list, axis=1)
    
        # FIX: Drop rows with all NaNs, but don't forward-fill
        # Forward-filling creates artificial price continuity
        adj_close_df = adj_close_df.dropna(how='all')

    except Exception as e:
        print(f"Data Acquisition Error: {e}")
        monthly_returns = pd.DataFrame()

    if not adj_close_df.empty:
        monthly_returns = adj_close_df.pct_change(fill_method=None)
    
        # FIX: Drop rows with all NaNs first
        monthly_returns = monthly_returns.dropna(how='all')
    
        # FIX: Filter columns with insufficient data (consistent with optimization codes)
        min_observations = 12  # At least 12 months of data
        monthly_returns = monthly_returns.loc[:, monthly_returns.notna().sum() >= min_observations]
    
        if monthly_returns.empty:
            print(f"No assets have at least {min_observations} observations.")
            monthly_returns = pd.DataFrame()
        else:
            # FIX: For statistics, we want per-asset analysis, so we keep individual series
            # We do NOT use dropna() for common intersection here, unlike optimization
            # Each asset is analyzed independently with its available data
            pass
    else:
        monthly_returns = pd.DataFrame()

    display_order = ['SP500', 'NDQ100', 'EEM', 'Gold Futures', 'AMD', 'AMZN', 'FORD', 'JPM', 'NVDA', 'PFE', 'TGT', 'XOM']
    available_cols = [col for col in display_order if col in monthly_returns.columns]
    monthly_returns = monthly_returns.reindex(columns=available_cols)

    # --- 2. Generate Histograms (First Slide) ---
    if not monthly_returns.empty:
        print("\n### Risk in Investments")
        print("\n**Histograms of various stock/index returns**")
        print(f"Assets with sufficient data: {', '.join(monthly_returns.columns)}\n")

        num_tickers = len(monthly_returns.columns)
        nrows = (num_tickers + 3) // 4
        fig, axes = plt.subplots(nrows, 4, figsize=(16, 4 * nrows))
        axes = axes.flatten()

        for i, ticker in enumerate(monthly_returns.columns):
            ax = axes[i]
            # Drop NaNs for each asset individually
            returns = monthly_returns[ticker].dropna()
        
            if len(returns) < 2:
                ax.text(0.5, 0.5, 'Insufficient Data', ha='center', va='center', transform=ax.transAxes)
                ax.set_xlabel(ticker)
                continue
        
            ax.hist(returns, bins=15, density=True, color='darkkhaki', alpha=0.8)
        
            mu, std = returns.mean(), returns.std()
            if std > 1e-9:  # Only plot normal curve if std is meaningful
                x = np.linspace(returns.min(), returns.max(), 100)
                p = norm.pdf(x, mu, std)
                ax.plot(x, p, linewidth=1.5, color='steelblue')
        
            ax.set_xlabel(ticker)
            ax.set_ylabel('Density')
            ax.grid(True, linestyle=':', alpha=0.6)

        for j in range(i + 1, len(axes)):
            axes[j].set_visible(False)

        plt.tight_layout()
        plt.show()

        print("\n---\n")

        # --- 3. Calculate and Display Summary Statistics (Second Slide) ---
        print("### Risk in Investments\n")

        summary_stats = pd.DataFrame(index=[
            'Mean', 'Standard Error', 'Median', 'Standard Deviation', 'Sample Variance',
            'CoV', 'Excess Kurtosis', 'Skewness', 'Range', 'Minimum', 'Maximum', 'Count'
        ])

        for ticker in monthly_returns.columns:
            # Drop NaNs for each asset individually
            returns = monthly_returns[ticker].dropna()
        
            if len(returns) < 2:
                # Not enough data for statistics
                summary_stats[ticker] = pd.Series({
                    'Mean': np.nan, 'Standard Error': np.nan, 'Median': np.nan,
                    'Standard Deviation': np.nan, 'Sample Variance': np.nan,
                    'CoV': np.nan, 'Excess Kurtosis': np.nan, 'Skewness': np.nan,
                    'Range': np.nan, 'Minimum': np.nan, 'Maximum': np.nan,
                    'Count': len(returns)
                })
                continue
        
            mean_val = returns.mean()
            std_val = returns.std()
        
            stats = {
                'Mean': mean_val,
                'Standard Error': returns.sem(),
                'Median': returns.median(),
                'Standard Deviation': std_val,
                'Sample Variance': returns.var(),
                'CoV': std_val / mean_val if abs(mean_val) > 1e-9 else np.nan,
                'Excess Kurtosis': kurtosis(returns, fisher=True, bias=False) if len(returns) >= 4 else np.nan,
                'Skewness': skew(returns, bias=False) if len(returns) >= 3 else np.nan,
                'Range': returns.max() - returns.min(),
                'Minimum': returns.min(),
                'Maximum': returns.max(),
                'Count': returns.count()
            }
            summary_stats[ticker] = pd.Series(stats)

        def format_percent(x):
            return f"{x:.2%}"

        formatters = {
            'Mean': format_percent,
            'Standard Error': format_percent,
            'Median': format_percent,
            'Standard Deviation': format_percent,
            'Minimum': format_percent,
            'Maximum': format_percent,
            'Range': format_percent
        }

        summary_display = summary_stats.style.format(formatters, na_rep="-").format(
            subset=pd.IndexSlice[['Sample Variance', 'CoV', 'Excess Kurtosis', 'Skewness'], :],
            formatter="{:.4f}"
        ).format(
            subset=pd.IndexSlice[['Count'], :],
            formatter="{:.0f}"
        )

        display(summary_display)

    else:
        print("Could not generate output because no valid data was downloaded.")
distribution_start_box = widgets.Text(
    value=default_start_date.isoformat(), description='Start:'
)
distribution_through_box = widgets.Text(
    value=default_through_date.isoformat(), description='Through:'
)
distribution_run_button = widgets.Button(description='Download and run', button_style='primary')
distribution_output = widgets.Output()

def on_distribution_run(_):
    with distribution_output:
        clear_output(wait=True)
        try:
            start_date = datetime.date.fromisoformat(distribution_start_box.value.strip().replace('.', '-'))
            through_date = datetime.date.fromisoformat(distribution_through_box.value.strip().replace('.', '-'))
        except ValueError:
            print('Invalid date. Use YYYY-MM-DD or YYYY.MM.DD.')
            return
        run_distribution_analysis(start_date, through_date)

distribution_run_button.on_click(on_distribution_run)
display(widgets.VBox([
    widgets.HBox([distribution_start_box, distribution_through_box]),
    distribution_run_button,
    distribution_output
]))


### Risk in Investments

*   A distribution with no skew, such as a normal distribution, is symmetrical, with the mean, median, and mode all being equal.
    
    *   A positively skewed distribution is a type of data distribution where most of the data points are clustered towards the lower end of the spectrum, with a smaller number of outliers near the higher end.
    
    *   A negatively skewed distribution is a distribution where the majority of values are concentrated on the right side of the distribution graph, while the left tail is longer.
    
    *   Coefficient of variation (CoV): Quantifies the relative variability in relation to the mean.

---


### Risk in Investments

Continue with the Gamma Corp. example from cells 8 and 9. Variance and standard deviation measure the dispersion of its possible returns around the expected return.

$$
E(r) = \sum_s p(s) \times r(s)
$$
$$
\sigma^2 = \sum_s p(s) \times [r(s) - E(r)]^2 \quad \text{and} \quad \sigma = \sqrt{\sigma^2}
$$

| State | Probability | Return |
| :--- | ---: | ---: |
| Boom | 0.10 | 0.4000 |
| Growth | 0.20 | 0.2500 |
| Normal | 0.40 | 0.1000 |
| Recession | 0.20 | -0.0500 |
| Depression | 0.10 | -0.2000 |

$$
\begin{aligned}
E(r) &= (0.10)(0.40) + (0.20)(0.25) + (0.40)(0.10)\\
&\quad + (0.20)(-0.05) + (0.10)(-0.20)\\
&= 0.10 = 10.00\%.
\end{aligned}
$$

$$
\begin{aligned}
\sigma^2 &= 0.10(0.30)^2 + 0.20(0.15)^2 + 0.40(0)^2\\
&\quad + 0.20(-0.15)^2 + 0.10(-0.30)^2\\
&= 0.0270,\\[2pt]
\sigma &= \sqrt{0.0270} = 0.1643 = 16.43\%.
\end{aligned}
$$

| Scenario | Prob. ($p_s$) | Return ($r_s$) | Deviation ($r_s - E(r)$) | Squared Deviation | Weighted Sq. Dev. ($p_s \times [r_s - E(r)]^2$) |
| :--- | ---: | ---: | ---: | ---: | ---: |
| Boom | 10% | 40.00% | 0.3000 | 0.090000 | 0.009000 |
| Growth | 20% | 25.00% | 0.1500 | 0.022500 | 0.004500 |
| Normal | 40% | 10.00% | 0.0000 | 0.000000 | 0.000000 |
| Recession | 20% | -5.00% | -0.1500 | 0.022500 | 0.004500 |
| Depression | 10% | -20.00% | -0.3000 | 0.090000 | 0.009000 |
| Total | 100% | | | Variance ($\sigma^2$) | 0.027000 |


---

### Risk in Investments

Continue with Gamma Corp., where $P_0=50$ and $r_f=4\%$.

| Market state | $p(s)$ | $P_1$ | $D_1$ | $r(s)$ | $p(s)r(s)$ | $p(s)[r(s)-E(r)]^2$ | $r(s)-r_f$ | $p(s)[r(s)-r_f]$ | $p(s)[(r(s)-r_f)-E(r-r_f)]^2$ |
| :--- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| Boom | 10% | 68.00 | 2.00 | 40.00% | 4.00% | 0.0090 | 36.00% | 3.60% | 0.0090 |
| Growth | 20% | 61.00 | 1.50 | 25.00% | 5.00% | 0.0045 | 21.00% | 4.20% | 0.0045 |
| Normal | 40% | 54.00 | 1.00 | 10.00% | 4.00% | 0.0000 | 6.00% | 2.40% | 0.0000 |
| Recession | 20% | 47.00 | 0.50 | -5.00% | -1.00% | 0.0045 | -9.00% | -1.80% | 0.0045 |
| Depression | 10% | 40.00 | 0.00 | -20.00% | -2.00% | 0.0090 | -24.00% | -2.40% | 0.0090 |
| Total | 100% | | | | $E(r)=10.00\%$ | $\sigma^2=0.0270$ | | $E(r-r_f)=6.00\%$ | $\sigma^2=0.0270$ |

The next code cell applies the same calculations to a separate five-state example for Delta Corp.

---


In [2]:
import numpy as np

# Separate five-state example: Delta Corp.
probs = [0.15, 0.30, 0.30, 0.20, 0.05]
returns = [0.44, 0.23, 0.06, -0.09, -0.40]
rf = 0.04
P0 = 50
prices = [70, 60, 52, 45, 30]
dividends = [2.00, 1.50, 1.00, 0.50, 0.00]

# Total Returns
print("Alternative example: Delta Corp.")
total_returns = [(p1 - P0 + d) / P0 for p1, d in zip(prices, dividends)]
print("Total Returns:", [f"{r*100:.2f}%" for r in total_returns])

# Expected Return
E_r = sum(p * r for p, r in zip(probs, returns))
print(f"E(r): {E_r * 100:.2f}%")

# Variance
var = sum(p * (r - E_r)**2 for p, r in zip(probs, returns))
print(f"Variance: {var:.6f}")

# Standard Deviation
std = np.sqrt(var)
print(f"Std Dev: {std * 100:.2f}%")

# Excess Returns
excess_returns = [r - rf for r in returns]
print("Excess Returns:", [f"{er*100:.2f}%" for er in excess_returns])

# Expected Excess Return
E_excess = sum(p * er for p, er in zip(probs, excess_returns))
print(f"E(r) - rf: {E_excess * 100:.2f}%")

# Variance of Excess Return
var_excess = sum(p * (er - E_excess)**2 for p, er in zip(probs, excess_returns))
print(f"Excess Variance: {var_excess:.6f}")

Alternative example: Delta Corp.
Total Returns: ['44.00%', '23.00%', '6.00%', '-9.00%', '-40.00%']
E(r): 11.50%
Variance: 0.042385
Std Dev: 20.59%
Excess Returns: ['40.00%', '19.00%', '2.00%', '-13.00%', '-44.00%']
E(r) - rf: 7.50%
Excess Variance: 0.042385


---

### Risk in Investments: Historical Average Return vs Historical Volatility
<center>
<img src = "../images/slide_4/pic_6.png" width="960">
</center>
---



In [3]:
# If needed, uncomment and run once:
# %pip install -U yfinance curl_cffi pandas numpy matplotlib ipywidgets

import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import ipywidgets as widgets
from IPython.display import display, HTML, clear_output
import datetime

# --- 1. Data Download and Preparation ---
tickers = ['^SP500TR', 'QQQ', 'EEM', 'GC=F', 'AMD', 'AMZN', 'F', 'JPM', 'NVDA', 'PFE', 'TGT', 'XOM']
ticker_map = {
    '^SP500TR': 'SP500', 'QQQ': 'NDQ100', 'EEM': 'EEM', 'GC=F': 'Gold Futures',
    'AMD': 'AMD', 'AMZN': 'AMZN', 'F': 'FORD', 'JPM': 'JPM',
    'NVDA': 'NVDA', 'PFE': 'PFE', 'TGT': 'TGT', 'XOM': 'XOM'
}

default_start_date = datetime.date(1999, 12, 31)
default_split_date = datetime.date(2010, 3, 31)
last_month_end = (pd.Timestamp.today().normalize().replace(day=1) - pd.Timedelta(days=1)).date()

def run_risk_return_analysis(start_date, through_date, split_date):
    if start_date is None or through_date is None or split_date is None:
        print('Select all three dates.')
        return
    if not start_date < split_date < through_date:
        print('The split date must fall between the start and through dates.')
        return
    end_date = through_date + datetime.timedelta(days=1)  # yfinance end is exclusive

    print(f"Downloading data for specified tickers from {start_date} through {through_date}...")

    try:
        data_downloaded = yf.download(
            tickers, start=start_date, end=end_date, interval='1mo',
            group_by='ticker', progress=False, auto_adjust=True, actions=False
        )
        if data_downloaded.empty:
            raise ValueError("yf.download returned an empty DataFrame.")
    
        # With auto_adjust=True, Close is the adjusted-close series.
        df_list = []
        for ticker in tickers:
            if ticker in data_downloaded.columns.get_level_values(0):
                ticker_data = data_downloaded[ticker]
                if 'Close' in ticker_data.columns and not ticker_data['Close'].isnull().all():
                    df_list.append(ticker_data['Close'].rename(ticker_map[ticker]))

        if not df_list:
            raise ValueError("No valid adjusted-close data extracted for any ticker.")
    
        adj_close_df = pd.concat(df_list, axis=1)
    
        # FIX: Don't forward-fill, just drop rows with all NaNs
        adj_close_df = adj_close_df.dropna(how='all')

    except Exception as e:
        print(f"Data Acquisition Error: {e}")
        adj_close_df = pd.DataFrame()

    if not adj_close_df.empty:
        monthly_returns = adj_close_df.pct_change(fill_method=None)
    
        # FIX: Drop rows with all NaNs
        monthly_returns = monthly_returns.dropna(how='all')
    
        # FIX: Filter columns with insufficient data
        min_observations = 12
        monthly_returns = monthly_returns.loc[:, monthly_returns.notna().sum() >= min_observations]
    
        if monthly_returns.empty:
            print(f"No assets have at least {min_observations} observations.")
            monthly_returns = pd.DataFrame()
        else:
            display_order = ['SP500', 'NDQ100', 'EEM', 'Gold Futures', 'AMD', 'AMZN', 'FORD', 'JPM', 'NVDA', 'PFE', 'TGT', 'XOM']
            available_cols = [col for col in display_order if col in monthly_returns.columns]
            monthly_returns = monthly_returns.reindex(columns=available_cols)
        
            print(f"Assets with sufficient data: {', '.join(monthly_returns.columns)}\n")
    else:
        monthly_returns = pd.DataFrame()

    # --- 2. Define Sample Periods ---
    if not monthly_returns.empty:
        full_sample_start = datetime.datetime.combine(start_date, datetime.time.min)
        full_sample_end = datetime.datetime.combine(through_date, datetime.time.max)
        sub_sample_1_start = full_sample_start
        sub_sample_1_end = datetime.datetime.combine(split_date, datetime.time.max)
        sub_sample_2_start = datetime.datetime.combine(
            split_date + datetime.timedelta(days=1), datetime.time.min
        )
        sub_sample_2_end = datetime.datetime.combine(through_date, datetime.time.max)

        # Slice the data according to the defined periods
        full_sample_data = monthly_returns.loc[full_sample_start:full_sample_end]
        sub_sample_1_data = monthly_returns.loc[sub_sample_1_start:sub_sample_1_end]
        sub_sample_2_data = monthly_returns.loc[sub_sample_2_start:sub_sample_2_end]
    
        # FIX: Check if subsamples are empty
        samples = {
            'Full Sample': (full_sample_data, full_sample_start, full_sample_end),
            'Sub-sample 1': (sub_sample_1_data, sub_sample_1_start, sub_sample_1_end),
            'Sub-sample 2': (sub_sample_2_data, sub_sample_2_start, sub_sample_2_end)
        }
    
        for name, (data, start, end) in samples.items():
            if data.empty:
                print(f"Warning: {name} ({start:%Y-%m-%d} to {end:%Y-%m-%d}) has no data.")

    # --- 3. Define Plotting Function ---
    def create_risk_return_plot(ax, data, title):
        if data.empty:
            ax.text(0.5, 0.5, 'No Data Available', ha='center', va='center', 
                    transform=ax.transAxes, fontsize=14)
            ax.set_title(title, fontsize=12)
            return
    
        # FIX: Drop NaNs per column for mean/std calculation
        # For risk-return, we analyze each asset independently
        means = data.mean()
        stds = data.std()
    
        # Filter out assets with insufficient data in this period
        valid_assets = means.notna() & stds.notna() & (stds > 0)
        means = means[valid_assets]
        stds = stds[valid_assets]
    
        if means.empty:
            ax.text(0.5, 0.5, 'Insufficient Data', ha='center', va='center',
                    transform=ax.transAxes, fontsize=14)
            ax.set_title(title, fontsize=12)
            return
    
        risk_return = pd.DataFrame({'Return': means, 'Risk': stds})
        ax.scatter(risk_return['Risk'], risk_return['Return'], color='#004B87', s=50)
    
        for ticker, row in risk_return.iterrows():
            ax.text(row['Risk'] * 1.02, row['Return'], ticker, fontsize=10, 
                    verticalalignment='center')
    
        ax.set_xlabel('Risk (Standard Deviation)', fontsize=10)
        ax.set_ylabel('Expected Return (Mean)', fontsize=10)
        ax.set_title(title, fontsize=12)
        ax.grid(True, linestyle='-', alpha=0.3)
        ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{x:.2%}'))
        ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda y, _: f'{y:.2%}'))

    # --- 4. Generate the Three Plots ---
    if not monthly_returns.empty:
        # --- Plot 1: Full Sample ---
        fig1, ax1 = plt.subplots(1, 1, figsize=(10, 6))
        title1 = (
            f"Risk & Return (Monthly, {start_date:%Y-%m-%d} to "
            f"{through_date:%Y-%m-%d})"
        )
        create_risk_return_plot(ax1, full_sample_data, title1)
        plt.tight_layout()
        plt.show()

        # --- Plot 2 & 3: Sub-samples ---
        fig2, (ax2, ax3) = plt.subplots(1, 2, figsize=(18, 7))
    
        title2 = (
            f"Risk & Return (Monthly, {start_date:%Y-%m-%d} to "
            f"{split_date:%Y-%m-%d})"
        )
        create_risk_return_plot(ax2, sub_sample_1_data, title2)
    
        second_period_start = split_date + datetime.timedelta(days=1)
        title3 = (
            f"Risk & Return (Monthly, {second_period_start:%Y-%m-%d} to "
            f"{through_date:%Y-%m-%d})"
        )
        create_risk_return_plot(ax3, sub_sample_2_data, title3)
    
        plt.tight_layout()
        plt.show()
    else:
        print("Could not generate plots because no valid data was downloaded.")
risk_return_start_box = widgets.Text(
    value=default_start_date.isoformat(), description='Start:'
)
risk_return_through_box = widgets.Text(
    value=last_month_end.isoformat(), description='Through:'
)
risk_return_split_box = widgets.Text(
    value=default_split_date.isoformat(), description='Split:'
)
risk_return_run_button = widgets.Button(description='Download and run', button_style='primary')
risk_return_output = widgets.Output()

def on_risk_return_run(_):
    with risk_return_output:
        clear_output(wait=True)
        try:
            start_date = datetime.date.fromisoformat(risk_return_start_box.value.strip().replace('.', '-'))
            through_date = datetime.date.fromisoformat(risk_return_through_box.value.strip().replace('.', '-'))
            split_date = datetime.date.fromisoformat(risk_return_split_box.value.strip().replace('.', '-'))
        except ValueError:
            print('Invalid date. Use YYYY-MM-DD or YYYY.MM.DD.')
            return
        run_risk_return_analysis(start_date, through_date, split_date)

risk_return_run_button.on_click(on_risk_return_run)
display(widgets.VBox([
    widgets.HBox([risk_return_start_box, risk_return_split_box, risk_return_through_box]),
    risk_return_run_button,
    risk_return_output
]))


---

### Risk in Investments
*   True means and variances are unobservable
    *   Possible scenarios like the ones used in the examples are unknown
*   Means and variances must therefore be estimated from historical observations
*   Expected Return Across States
    $$E(r)=\sum_{s=1}^{S}p(s)r(s)$$
*   Historical Arithmetic Average
    $$\bar r=\frac{1}{n}\sum_{t=1}^{n}r_t$$
    *   This is the equally weighted sample counterpart of expected return
*   Geometric (Time-Weighted) Average
    *   Terminal value of the investment:
        $$TV_n=(1+r_1)(1+r_2)\cdots(1+r_n)$$
    *   Geometric average:
        $$g=TV_n^{1/n}-1$$

---


### Risk in Investments
*   Sample Variance
    *   An unbiased estimator of the return variance under the usual assumptions:
        $$s^2 = \frac{1}{n-1}\sum_{t=1}^{n}[r_t-\bar r]^2$$
*   Sample Standard Deviation
    $$s=\sqrt{s^2}$$
*   Sharpe Ratio
    $$\frac{\text{Risk premium}}{\text{SD of excess returns}}$$

---


### Risk in Investments
*   Symmetric Returns
    *   Standard deviation is a good measure of risk
    *   If returns are jointly normal, linear portfolio returns are also symmetric
    *   Under jointly normal returns, the mean vector and covariance matrix fully characterize portfolio returns
    *   Under elliptical/normal returns, the correlation matrix summarizes dependence; otherwise correlations can miss tail dependence or asymmetry
*   What if excess returns are not normally distributed?
    *   STD is no longer a complete measure of risk
    *   Skewness:
        $$
        \text{Skew} = \text{Average} \left[\frac{(R - \bar{R})^3}{\hat{\sigma}^3}\right]
        $$
    *   Excess kurtosis:
        $$
        \text{Kurtosis} = \text{Average} \left[\frac{(R - \bar{R})^4}{\hat{\sigma}^4}\right] - 3
        $$
    *   Sharpe ratio is not a complete measure of portfolio performance

---



### Risk in Investments

*   The normal distribution is a useful benchmark, but many assets and strategies exhibit skewness and fat tails

*   Skewness and kurtosis
    *   Negative skewness indicates a longer or heavier left tail
    *   Positive excess kurtosis indicates more extreme outcomes than under a normal distribution

*   Implications for VaR and Expected Shortfall (ES/CVaR)
    *   Under normality, mean and standard deviation determine both VaR and ES
    *   Negative skewness and positive excess kurtosis can make normal-based VaR and ES understate downside risk
    *   ES is more sensitive to fat tails because it reflects the magnitude of losses beyond the VaR cutoff

---

### Risk and Risk Aversion
*   Speculation
    *   Taking a considerable risk for a commensurate gain
    *   Parties have heterogeneous expectations
*   Gambling
    *   Bet on an uncertain outcome for enjoyment
    *   Parties assign the same probabilities to the possible outcomes
*   Utility Values
    *   Investors are willing to consider:
        *   Risk-free assets
        *   Speculative positions with positive risk premiums
    *   Portfolio attractiveness
        *   Increases with expected return
        *   Decreases with risk
        *   What happens when return increases with risk?

---



### Risk and Risk Aversion
*   Utility Function
    $$U = E(r) - \frac{1}{2}A\sigma^2$$
    *   U = Utility
    *   E(r) = Expected return on the asset or portfolio
    *   A = Coefficient of risk aversion
    *   $\sigma^2$ = Variance of returns
    *   $1/2$ = A scaling factor
*   Risk Averse Investors: $\qquad A > 0$

*   Risk-Neutral Investors: $\quad \; \, A = 0$

*   Risk Lovers: $\qquad \qquad \quad \; A < 0$

---



### Risk and Risk Aversion

$$
U = E(r) - \frac{1}{2}A\sigma^2
$$

Assume the risk-free rate is 3% and an investor chooses among three portfolios:

| Portfolio | Risk premium | $E(r)$ | $\sigma$ |
| :--- | ---: | ---: | ---: |
| Alpha (low risk) | 3% | 6% | 8% |
| Beta (medium risk) | 7% | 10% | 15% |
| Gamma (high risk) | 12% | 15% | 25% |

| Risk aversion $A$ | $U_{\text{Alpha}}$ | $U_{\text{Beta}}$ | $U_{\text{Gamma}}$ | Highest utility |
| ---: | ---: | ---: | ---: | :--- |
| 2 | $0.06-\frac{1}{2}(2)(0.08^2)=0.0536$ | $0.10-\frac{1}{2}(2)(0.15^2)=0.0775$ | $0.15-\frac{1}{2}(2)(0.25^2)=0.0875$ | Gamma |
| 4 | $0.06-\frac{1}{2}(4)(0.08^2)=0.0472$ | $0.10-\frac{1}{2}(4)(0.15^2)=0.0550$ | $0.15-\frac{1}{2}(4)(0.25^2)=0.0250$ | Beta |
| 6 | $0.06-\frac{1}{2}(6)(0.08^2)=0.0408$ | $0.10-\frac{1}{2}(6)(0.15^2)=0.0325$ | $0.15-\frac{1}{2}(6)(0.25^2)=-0.0375$ | Alpha |

As risk aversion increases, the preferred portfolio shifts toward lower risk.

---

### Risk and Risk Aversion
*   Mean-Variance (M-V) Criterion
    *   Portfolio Alpha dominates Portfolio Beta if:
        $E(r_{\text{Alpha}}) \ge E(r_{\text{Beta}})$ and $\sigma_{\text{Alpha}} \le \sigma_{\text{Beta}}$ and at least one inequality is strict
*   Indifference Curves
    *   Equally preferred portfolios will lie in the mean-standard deviation plane on an indifference curve
    *   All portfolio points with the same utility value


<center>
<img src="../images/slide_4/pic_10.png" width="600"/>
</center>

---



### Portfolio Return
*   Portfolio return is the weighted average of the assets' returns, using beginning-of-period portfolio weights:
    $$
    r_p = \sum_{n=1}^N w_nr_n.
    $$

An investor holds the following portfolio at the beginning of the period:

| Asset | Price | Units | Initial value | Weight | Total return | Return contribution | Ending value |
| :--- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| Alpha | 50.00 | 80 | 4,000.00 | 40% | 10% | 4.00% | 4,400.00 |
| Beta | 100.00 | 30 | 3,000.00 | 30% | 5% | 1.50% | 3,150.00 |
| Gamma | 75.00 | 40 | 3,000.00 | 30% | -4% | -1.20% | 2,880.00 |
| Total | | | 10,000.00 | 100% | | 4.30% | 10,430.00 |

$$
r_p=(0.40)(0.10)+(0.30)(0.05)+(0.30)(-0.04)=0.043=4.30\%.
$$

The change in total portfolio value gives the same return:

$$
r_p=\frac{10{,}430-10{,}000}{10{,}000}=4.30\%.
$$

---



### Portfolio Expected Return

Continue with the Alpha, Beta, and Gamma portfolio from cell 30, with weights of 40%, 30%, and 30%.

$$
E(r_p)=\sum_{n=1}^N w_nE(r_n).
$$

| Market state | Probability | Alpha return | Beta return | Gamma return | Portfolio return |
| :--- | ---: | ---: | ---: | ---: | ---: |
| Boom | 20% | 20% | 12% | 18% | 17.00% |
| Growth | 30% | 14% | 8% | 10% | 11.00% |
| Normal | 35% | 10% | 5% | -4% | 4.30% |
| Recession | 15% | -8% | -2% | -10% | -6.80% |
| Expected return | | 10.50% | 6.25% | 3.70% | 7.19% |

The Normal state uses the same asset returns as cell 30 and therefore reproduces its 4.30% portfolio return.

$$
\begin{aligned}
E(r_p)
&=0.40(0.1050)+0.30(0.0625)+0.30(0.0370)\\
&=0.07185\approx 7.19\%.
\end{aligned}
$$

---


In [4]:
import numpy as np
import pandas as pd
from IPython.display import display

states = ['Boom', 'Growth', 'Normal', 'Recession']
probabilities = pd.Series([0.20, 0.30, 0.35, 0.15], index=states, name='Probability')

asset_returns = pd.DataFrame({
    'Alpha': [0.20, 0.14, 0.10, -0.08],
    'Beta': [0.12, 0.08, 0.05, -0.02],
    'Gamma': [0.18, 0.10, -0.04, -0.10]
}, index=states)

weights = pd.Series({'Alpha': 0.40, 'Beta': 0.30, 'Gamma': 0.30})

# Method 1: weight the assets' expected returns.
expected_asset_returns = asset_returns.mul(probabilities, axis=0).sum()
expected_by_assets = expected_asset_returns @ weights

# Method 2: calculate the portfolio return in each state, then take its expectation.
portfolio_returns = asset_returns @ weights
expected_by_states = probabilities @ portfolio_returns

state_table = pd.concat(
    [probabilities, asset_returns, portfolio_returns.rename('Portfolio')], axis=1
)
display(state_table.style.format('{:.2%}'))

print('Expected asset returns:')
for asset, expected_return in expected_asset_returns.items():
    print(f'  {asset}: {expected_return:.2%}')

print(f'\nExpected portfolio return from asset averages: {expected_by_assets:.3%}')
print(f'Expected portfolio return from state returns:  {expected_by_states:.3%}')

assert np.isclose(expected_by_assets, expected_by_states)


,Probability,Alpha,Beta,Gamma,Portfolio
Boom,20.00%,20.00%,12.00%,18.00%,17.00%
Growth,30.00%,14.00%,8.00%,10.00%,11.00%
Normal,35.00%,10.00%,5.00%,-4.00%,4.30%
Recession,15.00%,-8.00%,-2.00%,-10.00%,-6.80%


Expected asset returns:
  Alpha: 10.50%
  Beta: 6.25%
  Gamma: 3.70%

Expected portfolio return from asset averages: 7.185%
Expected portfolio return from state returns:  7.185%


### Portfolio Expected Return: Two Equivalent Methods

*   Asset-first method: calculate each asset's expected return and then apply the portfolio weights
*   State-first method: calculate the portfolio return in each state and then apply the state probabilities
*   Both methods give the same result because portfolio return is linear:

$$
E(r_p)=\sum_n w_nE(r_n)=\sum_s p(s)r_p(s)=7.185\%\approx 7.19\%.
$$

Cell 30 reports the realized return for one outcome; cells 31–32 calculate the expected return before the market state is known.

---


### Portfolio Risk

Continue with Alpha and Beta from the state-based example in cell 31.

$$
\sigma^2(r_i) = p_1[r_1-E(r_i)]^2 + p_2[r_2-E(r_i)]^2 + \cdots + p_n[r_n-E(r_i)]^2
$$
$$
\sigma_i = \sqrt{\sigma^2(r_i)}
$$

| Market state | Prob. | Alpha ($r_{\mathrm{Alpha}}$) | $p[r_{\mathrm{Alpha}}-E(r_{\mathrm{Alpha}})]^2$ | Beta ($r_{\mathrm{Beta}}$) | $p[r_{\mathrm{Beta}}-E(r_{\mathrm{Beta}})]^2$ |
| :--- | ---: | ---: | ---: | ---: | ---: |
| Boom | 20% | 20.0% | 0.001805 | 12.0% | 0.000661 |
| Growth | 30% | 14.0% | 0.000368 | 8.0% | 0.000092 |
| Normal | 35% | 10.0% | 0.000009 | 5.0% | 0.000055 |
| Recession | 15% | -8.0% | 0.005134 | -2.0% | 0.001021 |
| | | $E(r)=10.50\%$ | $\sigma^2_{\mathrm{Alpha}}=0.007315$ | $E(r)=6.25\%$ | $\sigma^2_{\mathrm{Beta}}=0.001829$ |
| | | | $\sigma_{\mathrm{Alpha}}=8.55\%$ | | $\sigma_{\mathrm{Beta}}=4.28\%$ |

---


### Portfolio Risk

*   When calculating the expected return of the portfolio, we take the weighted average of the expected returns of the assets.

*   We cannot use the same method for variance or standard deviation (i.e., simply taking a weighted average of individual asset variances is incorrect due to the effect of covariances).

*   **Covariance:**
    *   Measures the directional relationship between the returns of two assets.
    *   At the portfolio level, the sign and magnitude of the covariances between pairs of assets significantly impact the overall volatility (risk) of the portfolio.

*   For two assets, $i$ and $j$, when considering $N$ possible states of the world (scenarios), each with a probability $p_k$:

    $ \text{Cov}(r_i, r_j) = p_1 \cdot [r_{i,1} - E(r_i)] \cdot [r_{j,1} - E(r_j)] + p_2 \cdot [r_{i,2} - E(r_i)] \cdot [r_{j,2} - E(r_j)] + \dots + p_N \cdot [r_{i,N} - E(r_i)] \cdot [r_{j,N} - E(r_j)] $

    $ \text{Cov}(r_i, r_j) = \sum_{k=1}^{N} p_k \cdot [r_{i,k} - E(r_i)] \cdot [r_{j,k} - E(r_j)] $
    
    Where:
    *   $\text{Cov}(r_i, r_j)$ is the covariance between the returns of asset $i$ and asset $j$.
    *   $N$ is the number of possible states/scenarios and $p_k$ is the probability of state/scenario $k$ occurring.
    *   $r_{i,k}$ is the return of asset $i$ if state/scenario $k$ occurs and $E(r_i)$ is the expected return of asset $i$.
    *   $r_{j,k}$ is the return of asset $j$ if state/scenario $k$ occurs and $E(r_j)$ is the expected return of asset $j$.

---

### Portfolio Risk
*   Covariance
    *   A measure of the directional relationship between the returns on two assets.

$$
\operatorname{Cov}(r_{\mathrm{Alpha}},r_{\mathrm{Beta}})
=\sum_s p(s)[r_{\mathrm{Alpha}}(s)-E(r_{\mathrm{Alpha}})]
[r_{\mathrm{Beta}}(s)-E(r_{\mathrm{Beta}})]
$$

| Market state | Prob. | $r_{\mathrm{Alpha}}$ | $r_{\mathrm{Alpha}}-E(r_{\mathrm{Alpha}})$ | $r_{\mathrm{Beta}}$ | $r_{\mathrm{Beta}}-E(r_{\mathrm{Beta}})$ | Weighted cross-product |
| :--- | ---: | ---: | ---: | ---: | ---: | ---: |
| Boom | 20% | 20.0% | 0.0950 | 12.0% | 0.0575 | 0.001093 |
| Growth | 30% | 14.0% | 0.0350 | 8.0% | 0.0175 | 0.000184 |
| Normal | 35% | 10.0% | -0.0050 | 5.0% | -0.0125 | 0.000022 |
| Recession | 15% | -8.0% | -0.1850 | -2.0% | -0.0825 | 0.002289 |
| | | | | | | $\operatorname{Cov}(r_{\mathrm{Alpha}},r_{\mathrm{Beta}})=0.003588$ |

---


### Portfolio Risk
*   A concept closely related to covariance is the correlation.
    *   Covariance is not a standardized measure.
    *   Correlation is a function of covariance and expresses the relationship between the returns of two assets in terms of both direction and strength.
    *   The correlation is much easier to interpret since it is standardized and takes values between -1 and +1.
    $$
    \rho_{i,j} = \frac{\operatorname{Cov}(r_i,r_j)}{\sigma_i\sigma_j}
    $$
*   In our Alpha–Beta example,
    *   $\operatorname{Cov}(r_{\mathrm{Alpha}},r_{\mathrm{Beta}})=0.0035875$, $\sigma_{\mathrm{Alpha}}=8.55\%$, and $\sigma_{\mathrm{Beta}}=4.28\%$
    $$
    \rho_{\mathrm{Alpha},\mathrm{Beta}}
    =\frac{0.0035875}{0.085528\times0.042764}
    =0.9809
    $$

---


### Portfolio Risk
*   For a two-asset portfolio:
    $$
    \sigma^2(r_p)=w_{\mathrm{Alpha}}^2\sigma_{\mathrm{Alpha}}^2
    +w_{\mathrm{Beta}}^2\sigma_{\mathrm{Beta}}^2
    +2w_{\mathrm{Alpha}}w_{\mathrm{Beta}}\operatorname{Cov}(r_{\mathrm{Alpha}},r_{\mathrm{Beta}})
    $$

| $E(r_{\mathrm{Alpha}})$ | 10.50% | $E(r_{\mathrm{Beta}})$ | 6.25% | $\operatorname{Cov}(r_{\mathrm{Alpha}},r_{\mathrm{Beta}})$ | 0.003588 |
| :--- | :--- | :--- | :--- | :--- | :--- |
| $\sigma_{\mathrm{Alpha}}$ | 8.55% | $\sigma_{\mathrm{Beta}}$ | 4.28% | | |
| $\sigma_{\mathrm{Alpha}}^2$ | 0.007315 | $\sigma_{\mathrm{Beta}}^2$ | 0.001829 | | |

| $w_{\mathrm{Alpha}}$ | $w_{\mathrm{Beta}}$ | $E(r_p)$ | $\sigma_p^2$ | $\sigma_p$ |
| ---: | ---: | ---: | ---: | ---: |
| 100% | 0% | 10.500% | 0.007315 | 8.55% |
| 90% | 10% | 10.075% | 0.006589 | 8.12% |
| 80% | 20% | 9.650% | 0.005903 | 7.68% |
| 70% | 30% | 9.225% | 0.005256 | 7.25% |
| 60% | 40% | 8.800% | 0.004648 | 6.82% |
| 50% | 50% | 8.375% | 0.004080 | 6.39% |
| 40% | 60% | 7.950% | 0.003551 | 5.96% |
| 30% | 70% | 7.525% | 0.003061 | 5.53% |
| 20% | 80% | 7.100% | 0.002611 | 5.11% |
| 10% | 90% | 6.675% | 0.002200 | 4.69% |
| 0% | 100% | 6.250% | 0.001829 | 4.28% |

---


In [5]:
import numpy as np

# Alpha and Beta data from cell 31
probs = np.array([0.20, 0.30, 0.35, 0.15])
returns_alpha = np.array([0.20, 0.14, 0.10, -0.08])
returns_beta = np.array([0.12, 0.08, 0.05, -0.02])

E_alpha = probs @ returns_alpha
E_beta = probs @ returns_beta
var_alpha = probs @ (returns_alpha - E_alpha) ** 2
var_beta = probs @ (returns_beta - E_beta) ** 2
cov_alpha_beta = probs @ (
    (returns_alpha - E_alpha) * (returns_beta - E_beta)
)

# Two-asset portfolio calculations for different Alpha–Beta weights
weights = [(w_alpha, 1 - w_alpha) for w_alpha in np.linspace(1, 0, 11)]

print('w_Alpha  w_Beta  E(r_p)  Var(r_p)  Std(r_p)')
for w_alpha, w_beta in weights:
    E_rp = w_alpha * E_alpha + w_beta * E_beta
    var_p = (
        w_alpha**2 * var_alpha
        + w_beta**2 * var_beta
        + 2 * w_alpha * w_beta * cov_alpha_beta
    )
    std_p = np.sqrt(var_p)
    print(
        f'{w_alpha:7.0%} {w_beta:7.0%} '
        f'{E_rp:7.3%} {var_p:9.6f} {std_p:8.2%}'
    )


w_Alpha  w_Beta  E(r_p)  Var(r_p)  Std(r_p)
   100%      0% 10.500%  0.007315    8.55%
    90%     10% 10.075%  0.006589    8.12%
    80%     20%  9.650%  0.005903    7.68%
    70%     30%  9.225%  0.005256    7.25%
    60%     40%  8.800%  0.004648    6.82%
    50%     50%  8.375%  0.004080    6.39%
    40%     60%  7.950%  0.003551    5.96%
    30%     70%  7.525%  0.003061    5.53%
    20%     80%  7.100%  0.002611    5.11%
    10%     90%  6.675%  0.002200    4.69%
     0%    100%  6.250%  0.001829    4.28%


---

### Risk in Investments: Portfolio Risk

**Large portfolio variance**
$$ E(r_p) = \sum_{i=1}^{n} w_i E(r_i) \qquad \text{and} \qquad Var(r_p) = \sum_{i=1}^{n} \sum_{j=1}^{n} w_i w_j Cov(r_i, r_j) $$

**Diversification with an equally weighted portfolio**
$$ Var(r_p) = \frac{1}{n} (\text{Average Variance of the Individual Stocks}) + (1 - \frac{1}{n}) (\text{Average Covariance between the Stocks}) $$

<center>
<img src="../images/slide_4/large_portfolio_var.png" width="600"/>
</center>

---




In [6]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import ipywidgets as widgets
from IPython.display import display, clear_output

# --- Input Parameter Widgets ---
style_desc = {'description_width': '200px'}
widget_layout_left = widgets.Layout(width='350px')
widget_layout_right = widgets.Layout(width='350px')

# --- Defaults from the Alpha–Beta example ---
er_a_widget = widgets.FloatText(value=0.1050, step=0.0001, description='Expected Return Alpha:', style=style_desc, layout=widget_layout_left)
std_a_widget = widgets.FloatText(value=0.085528, step=0.0001, description='Std. Dev. Alpha:', style=style_desc, layout=widget_layout_left)
corr_ab_widget = widgets.FloatText(value=0.9809, step=0.0001, min=-1, max=1, description='Correlation (Alpha, Beta):', style=style_desc, layout=widget_layout_left)
cov_ab_widget = widgets.FloatText(value=0.0035875, step=0.0000001, description='Covariance (Alpha, Beta):', style=style_desc, layout=widget_layout_left)

er_b_widget = widgets.FloatText(value=0.0625, step=0.0001, description='Expected Return Beta:', style=style_desc, layout=widget_layout_right)
std_b_widget = widgets.FloatText(value=0.042764, step=0.0001, description='Std. Dev. Beta:', style=style_desc, layout=widget_layout_right)
# --- End of example defaults ---

# --- Output Widgets ---
output_gmvp = widgets.Output()
output_table = widgets.Output()
output_plot = widgets.Output()

# --- Recalculate Button ---
recalculate_button = widgets.Button(
    description="Recalculate & Plot",
    button_style='info',
    tooltip='Click to update frontiers and table',
    icon='refresh',
    layout=widgets.Layout(width='auto', min_width='150px', margin='10px 0 0 0')
)

# --- Linking Logic for Correlation and Covariance ---
updating_internally = False

def format_float_widget_value(widget, value, precision):
    try:
        if pd.isna(value):
            actual_value_to_set = 0.0
        else:
            actual_value_to_set = float(f"{value:.{precision}f}")
        current_widget_val = widget.value
        if current_widget_val is None or abs(current_widget_val - actual_value_to_set) > 10**-(precision+1):
            widget.value = actual_value_to_set
    except Exception:
        pass

def update_covariance(change):
    global updating_internally
    if updating_internally: return
    updating_internally = True
    try:
        std_a = std_a_widget.value; std_b = std_b_widget.value; corr_ab = corr_ab_widget.value
        if std_a is not None and std_b is not None and corr_ab is not None and std_a >= 0 and std_b >= 0:
            new_cov = corr_ab * std_a * std_b
            format_float_widget_value(cov_ab_widget, new_cov, 6)
    except Exception: pass
    finally: updating_internally = False

def update_correlation(change):
    global updating_internally
    if updating_internally: return
    updating_internally = True
    try:
        std_a = std_a_widget.value; std_b = std_b_widget.value; cov_ab = cov_ab_widget.value
        if std_a is not None and std_b is not None and cov_ab is not None:
            if std_a > 1e-9 and std_b > 1e-9:
                corr_val = cov_ab / (std_a * std_b)
                format_float_widget_value(corr_ab_widget, np.clip(corr_val, -1, 1), 4)
            elif abs(std_a) < 1e-9 or abs(std_b) < 1e-9:
                 format_float_widget_value(corr_ab_widget, 0.0 if abs(cov_ab) < 1e-9 else np.nan, 4)
            else:
                 format_float_widget_value(corr_ab_widget, np.nan, 4)
    except Exception: pass
    finally: updating_internally = False

def on_std_dev_change(change):
    update_covariance(None)

corr_ab_widget.observe(update_covariance, names='value')
cov_ab_widget.observe(update_correlation, names='value')
std_a_widget.observe(on_std_dev_change, names='value')
std_b_widget.observe(on_std_dev_change, names='value')

# --- Core Calculation and Plotting Logic ---
def calculate_and_display_frontiers(b=None):
    with output_gmvp: clear_output(wait=True)
    with output_table: clear_output(wait=True)
    with output_plot: clear_output(wait=True)
    try:
        er_a = er_a_widget.value; std_a = std_a_widget.value; var_a = std_a**2
        er_b = er_b_widget.value; std_b = std_b_widget.value; var_b = std_b**2
        cov_ab = cov_ab_widget.value

        if any(v is None for v in [er_a, std_a, er_b, std_b, cov_ab]):
             with output_plot: print("Error: One or more input values are missing or invalid.")
             return

        plot_weights_a = np.linspace(0, 1, 100)
        plot_portfolio_returns = np.array([w_a * er_a + (1 - w_a) * er_b for w_a in plot_weights_a])
        plot_portfolio_variances = np.array([(w_a**2 * var_a) + ((1 - w_a)**2 * var_b) + (2 * w_a * (1 - w_a) * cov_ab) for w_a in plot_weights_a])
        plot_portfolio_variances = np.maximum(0, plot_portfolio_variances)
        plot_portfolio_std_devs = np.sqrt(plot_portfolio_variances)

        denominator_gmvp = var_a + var_b - 2 * cov_ab
        w_a_gmvp = 0.5 if abs(denominator_gmvp) < 1e-9 else (var_b - cov_ab) / denominator_gmvp
        w_a_gmvp = np.clip(w_a_gmvp, 0, 1); w_b_gmvp = 1 - w_a_gmvp
        gmvp_return = w_a_gmvp * er_a + w_b_gmvp * er_b
        gmvp_variance = (w_a_gmvp**2 * var_a) + (w_b_gmvp**2 * var_b) + (2 * w_a_gmvp * w_b_gmvp * cov_ab)
        gmvp_variance = max(0, gmvp_variance)
        gmvp_std_dev = np.sqrt(gmvp_variance)

        with output_gmvp:
            display(widgets.HTML("<h4>Long-Only Global Minimum Variance Portfolio (GMVP):</h4>"))
            gmvp_df = pd.DataFrame({
                'Weight Alpha': [f"{w_a_gmvp*100:.2f}%"], 'Weight Beta': [f"{w_b_gmvp*100:.2f}%"],
                'Expected Return E(Rp)': [f"{gmvp_return*100:.3f}%"], 'Variance (σp²)': [f"{gmvp_variance:.6f}"],
                'Std. Deviation (σp)': [f"{gmvp_std_dev*100:.2f}%"]})
            display(gmvp_df)

        table_weights_a = np.array([1.0, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1, 0.0])
        table_data = []
        for w_a_t in table_weights_a:
            w_b_t = 1 - w_a_t
            rp_t = w_a_t * er_a + w_b_t * er_b
            var_p_t = (w_a_t**2 * var_a) + (w_b_t**2 * var_b) + (2 * w_a_t * w_b_t * cov_ab)
            var_p_t = max(0, var_p_t)
            std_p_t = np.sqrt(var_p_t)
            table_data.append([f"{w_a_t*100:.0f}%", f"{w_b_t*100:.0f}%", f"{rp_t*100:.3f}%", f"{var_p_t:.6f}", f"{std_p_t*100:.2f}%"])
        df_table = pd.DataFrame(table_data, columns=["w_Alpha", "w_Beta", "E(rp)", "σp²", "σp"])
        with output_table:
            display(widgets.HTML("<h4>Portfolio Combinations Table:</h4>"))
            display(df_table)

        with output_plot:
            fig, ax = plt.subplots(figsize=(9, 6))
            ax.plot(plot_portfolio_std_devs, plot_portfolio_returns, linestyle='-', color='blue', label='Long-Only Opportunity Set')
            
            sort_indices = np.argsort(plot_portfolio_std_devs)
            sorted_stds = plot_portfolio_std_devs[sort_indices]; sorted_rets = plot_portfolio_returns[sort_indices]
            gmvp_sorted_idx = np.argmin(sorted_stds)
            
            eff_stds = sorted_stds[gmvp_sorted_idx:]; eff_rets = sorted_rets[gmvp_sorted_idx:]
            
            final_eff_stds = [eff_stds[0]] if len(eff_stds)>0 else []; final_eff_rets = [eff_rets[0]] if len(eff_rets)>0 else []
            if len(eff_stds) > 0:
                for k in range(1, len(eff_stds)):
                    if eff_rets[k] >= final_eff_rets[-1] - 1e-6:
                        final_eff_stds.append(eff_stds[k]); final_eff_rets.append(eff_rets[k])
            
            if final_eff_stds:
                 ax.plot(final_eff_stds, final_eff_rets, linestyle='-', color='green', linewidth=2.5, label='Efficient Frontier', zorder=3)

            ax.scatter([std_a], [er_a], color='red', marker='o', s=100, label='Alpha', zorder=4)
            ax.scatter([std_b], [er_b], color='purple', marker='o', s=100, label='Beta', zorder=4)
            ax.scatter([gmvp_std_dev], [gmvp_return], color='black', marker='*', s=200, label='GMVP', zorder=5)
            ax.set_title('Long-Only Portfolios: Minimum Variance & Efficient Frontier'); ax.set_xlabel('Portfolio Std Dev (σp)'); ax.set_ylabel('Portfolio Exp Return (E(rp))')
            ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda y, _: f'{y*100:.2f}%')); ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{x*100:.2f}%'))
            ax.legend(); ax.grid(True, linestyle='--', alpha=0.7); plt.tight_layout(); plt.show()
    except Exception as e:
        with output_gmvp: clear_output(wait=True); print(f"Error in GMVP: {e}")
        with output_table: clear_output(wait=True); print(f"Error in Table: {e}")
        with output_plot: clear_output(wait=True); print(f"Error in Plot: {e}")

# --- Linking Button and Initial Display ---
recalculate_button.on_click(calculate_and_display_frontiers)

title_widget = widgets.HTML("<h2>Portfolio Risk Calculation: Alpha and Beta (Long-Only)</h2>", layout=widgets.Layout(width='100%'))
formula_cov_widget = widgets.HTML("For a two-asset portfolio (using Covariance):<br>    σ²(rp) = w₁²σ²(r₁) + w₂²σ²(r₂) + 2w₁w₂Cov(r₁, r₂)", layout=widgets.Layout(margin='0 0 2px 0'))
formula_corr_widget = widgets.HTML("For a two-asset portfolio (using Correlation):<br>    σ²(rp) = w₁²σ²(r₁) + w₂²σ²(r₂) + 2w₁w₂ρ₁₂σ₁σ₂", layout=widgets.Layout(margin='0 0 15px 0'))
inputs_col_a_corr_cov = widgets.VBox([er_a_widget, std_a_widget, corr_ab_widget, cov_ab_widget])
inputs_col_b = widgets.VBox([er_b_widget, std_b_widget], layout=widgets.Layout(margin='0 0 0 20px'))
left_panel_with_button = widgets.VBox([inputs_col_a_corr_cov, recalculate_button])
input_ui_grid = widgets.HBox([left_panel_with_button, inputs_col_b])

full_ui = widgets.VBox([
    title_widget, formula_cov_widget, formula_corr_widget, input_ui_grid,
    output_gmvp, widgets.HTML("<hr>"), output_table, widgets.HTML("<hr>"), output_plot
])

display(full_ui)
calculate_and_display_frontiers()

### Opportunity Set of Risky Assets
*   The minimum-variance frontier shows the risk–return opportunities available to the investor.

*   The efficient frontier contains the portfolio with the highest expected return at each risk level.

*   Capital allocation combines risky portfolio Alpha with a risk-free asset.
    *   If Gamma invests fraction $y$ in Alpha and $1-y$ in the risk-free asset:
        $$E(r_{\text{Gamma}})=yE(r_{\text{Alpha}})+(1-y)r_f$$
        $$\sigma_{\text{Gamma}}=y\sigma_{\text{Alpha}}, \qquad 0\leq y\leq1$$
    *   Changing $y$ determines Gamma's risk and expected return.

---


### Question 1

An investment analyst states: “For diversifying a stock portfolio, adding an asset with low correlation is good, but adding one with negative correlation is always better.”

Critically evaluate this statement. Describe a plausible scenario in which Alpha, with a correlation of $+0.10$ with the existing portfolio, offers a better mean–variance tradeoff than Beta, whose correlation is $-0.50$.

---


### Question 2

The mean–variance criterion states that Portfolio Alpha dominates Portfolio Beta if

$$
E(r_{\text{Alpha}}) \geq E(r_{\text{Beta}})
\quad \text{and} \quad
\sigma_{\text{Alpha}} \leq \sigma_{\text{Beta}},
$$

with at least one strict inequality.

If Alpha has both a higher expected return and higher risk than Beta, the criterion cannot rank them. How do the investor's utility function and indifference curves resolve this ambiguity and identify the optimal portfolio?

---


### Question 3

An analyst provides the following return forecasts for Alpha and Beta:

| Economic Scenario | Probability | Return on Alpha | Return on Beta |
| :---------------- | ----------: | --------------: | -------------: |
| Boom              | 20%         | 35%             | 15%            |
| Normal            | 50%         | 15%             | 9%             |
| Recession         | 30%         | -10%            | 3%             |

- Calculate the expected return and standard deviation of Alpha and Beta.

- Calculate the covariance and correlation between their returns.

- Calculate the expected return and standard deviation of a portfolio invested 40% in Alpha and 60% in Beta.

---


### Question 4

You are constructing a portfolio with two risky funds:

*   Alpha, a stock fund, has an expected return of 12% and a standard deviation of 18%.
*   Beta, a bond fund, has an expected return of 5% and a standard deviation of 7%.
*   The correlation between Alpha and Beta is $+0.15$.

a. Calculate the weights of Alpha and Beta in the global minimum variance portfolio (GMVP).

b. Calculate the GMVP's expected return and standard deviation.

---


### Question 5

Gamma is a complete portfolio that invests 70% in risky portfolio Alpha and 30% in a risk-free asset. Alpha has an expected return of 15% and a standard deviation of 22%, while the risk-free rate is 4%.

a. Calculate Gamma's expected return.

b. Calculate Gamma's standard deviation.

c. For an investor with a risk-aversion coefficient of $A=3$, calculate Gamma's utility. Would the investor prefer Gamma or portfolio Beta, which has an expected return of 11% and a standard deviation of 14%?

---


### What is next?
*   Portfolio Theory and Practice II
    *   Readings: Ch. 7
*   Suggested Problems
    *   Chapter 5: 7, 13
    *   Chapter 5 CFA Problems: 1, 2, 3, 4, 5, 6.
    *   Chapter 6: 4, 13, 14, 15, 16, 17, 18.
    *   Chapter 6 CFA Problems: 1, 2, 3, 8

---
